searchlib/src/vespa/searchlib/tensor/distance_calculator.h

```diff
@@ -36,27 +36,44 @@ class DistanceCalculator {
         assert(_query_tensor != nullptr);
         return *_query_tensor;
     }
-    const BoundDistanceFunction& function() const { return *_dist_fun; }
+    const BoundDistanceFunction& function() const noexcept { return *_dist_fun; }
+    bool has_single_subspace() const noexcept { return _attr_tensor.getTensorType().is_dense(); }
 
+    template<bool has_single_subspace=false>
     double calc_raw_score(uint32_t docid) const {
-        auto vectors = _attr_tensor.get_vectors(docid);
-        double result = _dist_fun->min_rawscore();
-        for (uint32_t i = 0; i < vectors.subspaces(); ++i) {
-            double distance = _dist_fun->calc(vectors.cells(i));
-            double score = _dist_fun->to_rawscore(distance);
-            result = std::max(result, score);
+        if (has_single_subspace) {
+            double distance = _dist_fun->calc(_attr_tensor.get_vector(docid, 0));
```

Comment:
Shouldn't we check for `cells.size == 0` here as well? If so, that case should be covered by a unit test.

Reply:
We should also use `max(_dist_fun->min_rawscore(), this_score)` to match existing behaviour.